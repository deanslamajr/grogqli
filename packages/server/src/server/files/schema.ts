import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
