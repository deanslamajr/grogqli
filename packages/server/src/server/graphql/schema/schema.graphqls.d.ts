/* 19a699f8a8761279db5b06650f17155319410b52
 * This file is automatically generated by graphql-let. */

import { GraphQLResolveInfo } from 'graphql';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
    [key: string]: unknown;
}> = {
    [K in keyof T]: T[K];
};
/** All built-in and custom scalars, mapped to their actual values */
export declare type Scalars = {
    ID: string;
    String: string;
    Boolean: boolean;
    Int: number;
    Float: number;
};
export declare type Recording = {
    __typename?: 'Recording';
    replaceThisBullshxt: Scalars['String'];
};
export declare type Query = {
    __typename?: 'Query';
    getRecordings: Array<Maybe<Recording>>;
};
export declare type ResolverTypeWrapper<T> = Promise<T> | T;
export declare type LegacyStitchingResolver<TResult, TParent, TContext, TArgs> = {
    fragment: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type NewStitchingResolver<TResult, TParent, TContext, TArgs> = {
    selectionSet: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type StitchingResolver<TResult, TParent, TContext, TArgs> = LegacyStitchingResolver<TResult, TParent, TContext, TArgs> | NewStitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | StitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type ResolverFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => Promise<TResult> | TResult;
export declare type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => AsyncIterator<TResult> | Promise<AsyncIterator<TResult>>;
export declare type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<{
        [key in TKey]: TResult;
    }, TParent, TContext, TArgs>;
    resolve?: SubscriptionResolveFn<TResult, {
        [key in TKey]: TResult;
    }, TContext, TArgs>;
}
export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
    resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}
export declare type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> = SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs> | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;
export declare type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> = ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>) | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;
export declare type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (parent: TParent, context: TContext, info: GraphQLResolveInfo) => Maybe<TTypes> | Promise<Maybe<TTypes>>;
export declare type IsTypeOfResolverFn<T = {}> = (obj: T, info: GraphQLResolveInfo) => boolean | Promise<boolean>;
export declare type NextResolverFn<T> = () => Promise<T>;
export declare type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (next: NextResolverFn<TResult>, parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
/** Mapping between all available schema types and the resolvers types */
export declare type ResolversTypes = {
    Recording: ResolverTypeWrapper<Recording>;
    String: ResolverTypeWrapper<Scalars['String']>;
    Query: ResolverTypeWrapper<{}>;
    Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
};
/** Mapping between all available schema types and the resolvers parents */
export declare type ResolversParentTypes = {
    Recording: Recording;
    String: Scalars['String'];
    Query: {};
    Boolean: Scalars['Boolean'];
};
export declare type RecordingResolvers<ContextType = any, ParentType extends ResolversParentTypes['Recording'] = ResolversParentTypes['Recording']> = {
    replaceThisBullshxt?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
    __isTypeOf?: IsTypeOfResolverFn<ParentType>;
};
export declare type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
    getRecordings?: Resolver<Array<Maybe<ResolversTypes['Recording']>>, ParentType, ContextType>;
};
export declare type Resolvers<ContextType = any> = {
    Recording?: RecordingResolvers<ContextType>;
    Query?: QueryResolvers<ContextType>;
};
/**
 * @deprecated
 * Use "Resolvers" root object instead. If you wish to get "IResolvers", add "typesPrefix: I" to your config.
 */
export declare type IResolvers<ContextType = any> = Resolvers<ContextType>;

          
// This is an extra code in addition to what graphql-codegen makes.
// Users are likely to use 'graphql-tag/loader' with 'graphql-tag/schema/loader'
// in webpack. This code enables the result to be typed.
import { DocumentNode } from 'graphql'
export default typeof DocumentNode
