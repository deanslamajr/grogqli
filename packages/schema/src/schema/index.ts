import { typeDefs as sharedTypes } from './shared';
import { typeDefs as queryTypes } from './queries';
import { typeDefs as mutationTypes } from './mutations';

export const typeDefs = [
  ...queryTypes,
  ...mutationTypes,
  ...sharedTypes,
];