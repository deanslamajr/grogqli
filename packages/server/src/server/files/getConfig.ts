import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
