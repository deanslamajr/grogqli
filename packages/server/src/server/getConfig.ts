import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}