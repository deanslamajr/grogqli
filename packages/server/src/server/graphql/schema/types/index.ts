import { join } from 'path'
import { loadFilesSync } from '@graphql-tools/load-files'
import { mergeTypeDefs } from '@graphql-tools/merge'

const loadedFiles = loadFilesSync(join(process.cwd(), '**/*.graphqls'))
export const typeDefs = mergeTypeDefs(loadedFiles)