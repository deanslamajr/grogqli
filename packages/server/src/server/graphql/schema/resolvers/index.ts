import {Resolvers} from '@grogqli/schema';

import { queryResolver } from './queries';
import {mutationResolver } from './mutations';
import {subscriptionResolver} from './subscriptions';

export const resolvers: Partial<Resolvers> = {
  Query: queryResolver,
  Mutation: mutationResolver,
  Subscription: subscriptionResolver
};
