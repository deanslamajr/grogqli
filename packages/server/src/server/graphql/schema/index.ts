import { makeExecutableSchema } from '@graphql-tools/schema';
import { typeDefs } from '@grogqli/schema';

import { resolvers } from './resolvers';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
