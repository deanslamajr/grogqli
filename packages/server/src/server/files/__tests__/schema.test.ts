import { SchemasMappingsInput } from '@grogqli/schema';
import path from 'path';

import testSchema from './schema_tests_grogqli/schemas/someSchemaId/schema.json';

import { getConfig } from '../getConfig';
import {
  NEW_SCHEMA_NAME,
  conditionallyCreateOrUpdateSchemaRecordings,
  SchemaDifferences,
} from '../schema';

jest.mock('../getConfig');

describe('conditionallyCreateOrUpdateSchemaRecordings', () => {
  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      return path.join(__dirname, 'schema_tests_grogqli');
    });
  });

  describe(`if a given schemaMapping.targetSchemaId === 'NEW'`, () => {
    describe('if schemaMapping.schemaName === null', () => {
      it('should throw', async () => {
        const schemaMappings: SchemasMappingsInput[] = [
          {
            opsRecordingsSchemaHash: '',
            opsRecordingsSchemaUrl: '',
            targetSchemaId: NEW_SCHEMA_NAME,
            schemaName: null,
          },
        ];

        await expect(
          conditionallyCreateOrUpdateSchemaRecordings(schemaMappings)
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe(`if a given schemaMapping.targetSchemaId !== 'NEW'`, () => {
    describe('if the given schemaMapping.opsRecordingsSchemaHash does not match the hash associated with the given schemaMapping.targetSchemaId', () => {
      describe('if there is a breaking change: TYPE_REMOVED', () => {
        it('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: TYPE_CHANGED_KIND', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: TYPE_REMOVED_FROM_UNION', () => {
        it('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: VALUE_REMOVED_FROM_ENUM', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: REQUIRED_INPUT_FIELD_ADDED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: IMPLEMENTED_INTERFACE_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: FIELD_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: FIELD_CHANGED_KIND', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: REQUIRED_ARG_ADDED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: ARG_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: ARG_CHANGED_KIND', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: DIRECTIVE_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: DIRECTIVE_ARG_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: REQUIRED_DIRECTIVE_ARG_ADDED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: DIRECTIVE_REPEATABLE_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });

      describe('if there is a breaking change: DIRECTIVE_LOCATION_REMOVED', () => {
        xit('should return info about the breaking change', async () => {
          const schemaMappings: SchemasMappingsInput[] = [
            {
              opsRecordingsSchemaHash: 'hashOfSomeSchemaWithATypeRemoved',
              opsRecordingsSchemaUrl: '',
              targetSchemaId: 'someSchemaId',
              schemaName: null,
            },
          ];

          const actual = await conditionallyCreateOrUpdateSchemaRecordings(
            schemaMappings
          );

          expect(actual).toHaveLength(1);
          expect(actual[0]).toHaveProperty('breakingChanges');
          expect(actual).toMatchSnapshot();
        });
      });
    });

    describe('if the given schemaMapping.opsRecordingsSchemaHash matches the hash associated with the given schemaMapping.targetSchemaId', () => {
      it('should resolve the expected data', async () => {
        const schemaMappings: SchemasMappingsInput[] = [
          {
            opsRecordingsSchemaHash: testSchema.hash,
            opsRecordingsSchemaUrl: '',
            targetSchemaId: 'someSchemaId',
            schemaName: null,
          },
        ];

        await expect(
          conditionallyCreateOrUpdateSchemaRecordings(schemaMappings)
        ).resolves.toMatchSnapshot();
      });
    });
  });
});
