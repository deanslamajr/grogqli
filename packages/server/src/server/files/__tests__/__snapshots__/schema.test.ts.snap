// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditionallyCreateOrUpdateSchemaRecordings if a given schemaMapping.targetSchemaId !== 'NEW' if the given schemaMapping.opsRecordingsSchemaHash does not match the hash associated with the given schemaMapping.targetSchemaId testing particular schema changes if a type\`s field is changed to target a different type should detect FIELD_CHANGED_KIND 1`] = `
Array [
  Object {
    "description": "Query.anotherThing changed type from AnotherThing to YetAnotherThing.",
    "type": "FIELD_CHANGED_KIND",
  },
]
`;

exports[`conditionallyCreateOrUpdateSchemaRecordings if a given schemaMapping.targetSchemaId !== 'NEW' if the given schemaMapping.opsRecordingsSchemaHash matches the hash associated with the given schemaMapping.targetSchemaId should resolve the expected data 1`] = `
Array [
  Object {
    "schemaHash": "booBopImASchemaHaSh",
    "schemaId": "someSchemaId",
  },
]
`;

exports[`conditionallyCreateOrUpdateSchemaRecordings if a given schemaMapping.targetSchemaId === 'NEW' if schemaMapping.schemaName === null should throw 1`] = `"\\"input.schemaMappings.schemaName\\" is required for new schema creation."`;
